sql: still more routine dependency improvements

DrewKimball opened the pull request:
#### sql: still more routine dependency improvements

I missed in #158935 that check constraints and partial index predicates
are additional places where we unnecessarily track transitive column
dependencies when creating a routine. The fix is gated behind the same
session setting as #158935.

Fixes #158154

Release note (sql change): Fixed two cases where creation of a routine
resolves unnecessary column dependencies, which can prevent drop of the
column without first dropping the routine. Here, the unnecessary
dependencies are due to references within CHECK constraints, including
those for RLS policies and hash-sharded indexes, as well as those in
partial index predicates. The fix is gated behind the session setting
`use_improved_routine_deps_triggers_and_computed_cols`, which is off by
default before 26.1.

#### optbuilder: don't track deps for virtual tables

A previous fix in #146250 added dependency tracking for tables in
appendOrdinaryColumnsFromTable. This fix missed that the table could
be a virtual table, in which case we don't need (and in some places
can't handle) tracking dependencies. This commit adds a filter to
avoid tracking virtual tables.

The only place I've found where this mattered was in the cross-DB
check for views, which led to a test flake after some changes in
the previous commit.

Informs #158154

Release note: None

DrewKimball committed:
sql: still more routine dependency improvements

I missed in #158935 that check constraints and partial index predicates
are additional places where we unnecessarily track transitive column
dependencies when creating a routine. The fix is gated behind the same
session setting as #158935.

Fixes #158154

Release note (sql change): Fixed two cases where creation of a routine
resolves unnecessary column dependencies, which can prevent drop of the
column without first dropping the routine. Here, the unnecessary
dependencies are due to references within CHECK constraints, including
those for RLS policies and hash-sharded indexes, as well as those in
partial index predicates. The fix is gated behind the session setting
`use_improved_routine_deps_triggers_and_computed_cols`, which is off by
default before 26.1.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>  Nice catch!

Does this fix the case for enums, which also synthesize check constraints? Is it worth adding this test:

```sql
CREATE TYPE typ AS ENUM ('foo', 'bar');

CREATE TABLE t (a INT, b typ);

CREATE PROCEDURE p() LANGUAGE SQL AS $$
  INSERT INTO t (a) VALUES (1);
  UPDATE t SET a = a + 1;
$$;

CALL p();

ALTER TABLE t DROP COLUMN b;

CALL p();
```

We don't have to worry about partial indexes for now, because of https://github.com/cockroachdb/cockroach/issues/97813. I'll add a note to that issue pointing this out. If you think it's worth a TODO in the code where we build partial index predicates, then you can add one.

> @mgartner reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159126#-:-Og87138EWEdDhK2Y4Cu:bptflb7)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @ZhouXing19)



<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice find!! 

> @ZhouXing19 reviewed 4 of 4 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159126#-:-Og8KXiX4vR_mX5OCQDW:bskuupf)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball commented:
@mgartner good point about enums and partial indexes. I added an enum test (already fixed), but noticed that there were several more cases we didn't handle for partial indexes. Those are now fixed, with test cases added.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> @mgartner reviewed 5 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159126#-:-Og8gK_N6fImYegv4noF:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @ZhouXing19)

___
*[`pkg/sql/opt/optbuilder/partial_index.go` line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159126#-Og8f_i986Mt2EMLLM58:-Og8f_i986Mt2EMLLM59:b1l5zx) ([raw file](https://github.com/cockroachdb/cockroach/blob/caecd97f0490ab7118db222b2374275a7c1cf216/pkg/sql/opt/optbuilder/partial_index.go#L128)):*
> ```Go
> 		// dependencies on any UDTs or columns referenced by the predicate. The
> 		// partial index will already prevent dropping such UDTs or columns.
> 		defer b.DisableSchemaDepTracking()()
> ```

I'm wondering if this is the only addition you need for partial indexes. It looks like the three others you added all lead down to this method - the metadata and two arbiter cases. Did you try making just this one addition?


<!-- Sent from Reviewable.io -->

mgartner commented on pkg/sql/logictest/testdata/logic_test/row_level_security:
```diff
@@ -5930,4 +5930,61 @@ DROP TABLE accounts, non_rls;
+subtest regression_158154
+
+statement ok
+CREATE TABLE t158154 (c0 INT, c1 TEXT DEFAULT 'foobarbaz');
```
Tests are failing due to:

```
pkg/sql/logictest/testdata/logic_test/row_level_security:5964: 
        expected success, but found
        (0A000) unimplemented: multi-column-family check constraints are not yet supported under read committed isolation
        mutation_builder.go:1193: in addCheckConstraintCols()
```

Locking the column families for this table should do the trick.

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/159126)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner and @ZhouXing19)

___
*[`pkg/sql/opt/optbuilder/partial_index.go` line 128 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/159126#-Og8f_i986Mt2EMLLM58:-OgIaM5z-ReWYQb5md9n:b-c6k114) ([raw file](https://github.com/cockroachdb/cockroach/blob/caecd97f0490ab7118db222b2374275a7c1cf216/pkg/sql/opt/optbuilder/partial_index.go#L128)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

I'm wondering if this is the only addition you need for partial indexes. It looks like the three others you added all lead down to this method - the metadata and two arbiter cases. Did you try making just this one addition?
</blockquote></details>

I believe each of those cases ends up building the predicate directly with `buildScalar`, so it's necessary. I did try removing some of them individually to make sure the tests fail as expected, though it wasn't exhaustive.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
Unit tests flaked, just kicked off a retry.

DrewKimball committed:
optbuilder: don't track deps for virtual tables

A previous fix in #146250 added dependency tracking for tables in
appendOrdinaryColumnsFromTable. This fix missed that the table could
be a virtual table, in which case we don't need (and in some places
can't handle) tracking dependencies. This commit adds a filter to
avoid tracking virtual tables.

The only place I've found where this mattered was in the cross-DB
check for views, which led to a test flake after some changes in
the previous commit.

Informs #158154

Release note: None

DrewKimball commented:
blathers backport 26.1
